### ISSUE TITLE
Bounds on probabilities under measurement error

### ISSUE THREAD

**Issue by nsfinkelstein** [sachsmc/causaloptim#1]
Thanks for publishing this great library!

I am working on a related problem, and am wondering whether I can make use of your work on this library. Given a variable Y measured with error, I am interested in bounding P(Y=0). It seems this should be possible using a linear programming approach. The [website](https://sachsmc.shinyapps.io/causaloptimweb/) corresponding to this library says:

> The variables may be potential outcomes which are denoted by parentheses, and _they may also be observed outcomes which do not have parentheses_.

which would seem to indicate that this library can handle cases outside of causal queries.

Below is code for this question, adapted from your measurement error example.

```R
library(igraph)
library(causaloptim)

graph <- graph_from_literal(Ul -+ X -+ Y -+ Y2, Ur -+ Y, Ur -+ Y2)
V(graph)$leftside <- c(1, 1, 0, 0, 0)
V(graph)$latent <- c(1, 0, 1, 0, 1)
E(graph)$rlconnect <- c(0, 0, 0, 0, 0)
E(graph)$edge.monotone <- c(0, 0, 0, 0, 0)

analysis <- analyze_graph(graph, NULL, 'p{Y = 0}')

print(analysis)
```

This outputs, among other things:

```R
$objective
[1] "q1 + q2 + q3 + q4"

$q.vals
   Y Y2
1  0  0
2  1  0
3  2  0
4  3  0
5  0  1
6  1  1
7  2  1
8  3  1
9  0  2
10 1  2
11 2  2
12 3  2
13 0  3
14 1  3
15 2  3
16 3  3
```

However, I do not see how this stated objective maps to P(Y = 0). It even includes q4, for which Y is never equal to 0 (assuming you are using the same response function classes as Balke). 

It seems to me the objective should rather be something like the following (which is still linear in q):

```
P(X = 0)(q1 + q2 + q5 + q6 + q9 + q10 + q13 + q14) + P(X = 1)(q1 + q3 + q5 + q7 + q9 + q11 + q13 + q15)
```

In other words, `P(X = 0) * P(Y = 0 | X = 0) + P(X = 1) * P(Y = 0 | X = 1)`.

Have I misunderstood the output of the program, or how to use the software?

Thanks!

**Comment by sachsmc** [comment]
Hi Noam, 

That's a good question and an interesting use of the package. Theoretically it should work just fine, and it does provide bounds if you run `bnds <- optimize_effect(analysis)`. I think the source of confusion is the fact that the program performs some variable reduction on the qs before doing the optimization, therefore the qs that are output by objective do not actually correspond to the response function variable values in the q.vals table, but rather sums of them. From the output alone, it is not possible to determine what the relationship is between the qs and the reduced qs. 

In your example, when I "look under the hood", the objective in terms of the qs before the reduction step are: 

    q00 + q01 + q02 + q03 + q10 + q11 + q12 + q13 + q20 + q21 + q22 + q23

where the first index corresponds to the Y variable and the second to the Y2 variable. You can see that this does not include the cases where Y is always 1. 

Hopefully this clears up your question. I'm going to leave this issue open, and consider adding the unreduced qs to the output of  `analyze_graph` and also possibly the response functions. 

Best wishes,
Mike

**Comment by nsfinkelstein** [comment]
Thanks for the quick response! I wasn't aware that there was variable reduction happening in the analyze step. That helps clarify things.

The reason I was looking at the analysis step is that the bounds returned by the program didn't match an intuition I have about what they should be. My intuition could easily be wrong, but I do have one follow up question, to make sure I follow where these bounds are coming from.

In the objective you posted, all response function variables are summed except for those that always lead to Y = 1.  This seems to be suggesting that P(Y = 0) is the same as the probability that Y does not _always_ equal 1, i.e. `1 - P(Y(X = 0) = 1, Y(X = 1) = 1)`. However, that Y does not always equal 1 does not mean that it may not happen to equal 1 in the observed data.

In other words, it doesn't make sense to me that the full values of `\sum_j q1j` and `\sum_j q2j` should be added this objective. Instead it seems that these sums should be weighted by the proportion of the time they actually do lead to P(Y = 0). These weights are P(X = 0) and P(X= 1) respectively, as the "left side" must be randomized with respect to the "right side". This is where I got the objective I put in the original question, which could also be phrased in terms of the response function variables you used as follows:

```
q00 + q01 + q02 + q03 
  + P(X = 0)(q10 + q11 + q12 + q13)
  + P(X = 1)(q20 + q21 + q22 + q23)
```

Am I misunderstanding something about how the software is performing its calculations, or about the representation of the objective?

Thanks again,
Noam

**Comment by sachsmc** [comment]
Sorry, but your intuition is not correct. You can define the response functions variable for X, and the corresponding parameters, say k0 and k1. Then you can work out the objective in terms of the qs and the ks by evaluating the response functions. You will see that the ks can then be factored out of the full sum (the weights are not as you think they are), and hence do not need to factor in to the optimization. This is what is proven in general in Theorem 3 of the paper:

https://sachsmc.github.io/causaloptim/articles/CausalBoundsMethods.pdf

**Comment by nsfinkelstein** [comment]
Thanks for engaging about this. I would like to make sure I follow entirely before I use these results. You can feel free to disengage if you think I'm missing something.

I have tried to follow your description of the derivation of the objective function. In the final equation of section 3.3, I have the following for set for `\Gamma`, where each element is a set of response function variables, of the form `(ki, rjk)`, where `P(rjk) = qij`:

```
\Gamma = {
  (k0, r00), (k0, r01), (k0, r02), (k0, r03),
  (k0, r10), (k0, r11), (k0, r12), (k0, r13),
  (k1, r00), (k1, r01), (k1, r02), (k1, r03),
  (k1, r20), (k1, r21), (k1, r22), (k1, r23),
}
```

Then `Qv` is just the sum of the products corresponding to each of these tuples, as the probabilities of the ks and the rs are independent, so P(k1, r00) = P(k1)q00.

The P(k)s can be factored out of the sum for for the `q0k` variables, which would lead to the objective I mentioned, as `P(k0) + P(k1) = 1`. However I fail to see how the P(k) can be factored out for the other terms in the sum. I am sure Theorem 3 is correct; I'm just not sure `Qv` as described in the paper corresponds to the objective the software is using.

There is one other way to see that something strange may be going on. In the program below, I calculate bounds on P(Y = 0) and P(Y = 1). In both cases the software returns non-zero lower bounds (except when A is independent of Y2), but no upper bounds. However, this means the bounds cannot be sharp. If there is a non-zero lower bound on P(Y = 0), then there trivially is a non-one upper bound on P(Y = 1). In particular, P(Y = 1) cannot exceed one minus the lower bound of P(Y = 0); otherwise P(Y = 0) would have to be less than it's lower bound, as P(Y = 0) + P(Y = 1) = 1.

```R
library(igraph)
library(causaloptim)

graph <- graph_from_literal(Ul -+ X -+ Y -+ Y2, Ur -+ Y, Ur -+ Y2)
V(graph)$leftside <- c(1, 1, 0, 0, 0)
V(graph)$latent <- c(1, 0, 1, 0, 1)
E(graph)$rlconnect <- c(0, 0, 0, 0, 0)
E(graph)$edge.monotone <- c(0, 0, 0, 0, 0)

analysis <- analyze_graph(graph, NULL, 'p{Y = 0}')
solution = optimize_effect(analysis)
print('Bounds for P(Y = 0)')
print(solution)

analysis <- analyze_graph(graph, NULL, 'p{Y = 1}')
solution = optimize_effect(analysis)
print('Bounds for P(Y = 1)')
print(solution)
```

Output:

```
[1] "Bounds for P(Y = 0)"

MAX {
p0_0 - p0_1
- p0_0 + p0_1
}

 
MIN {
1
}

[1] "Bounds for P(Y = 1)"

MAX {
p0_0 - p0_1
- p0_0 + p0_1
}

 
MIN {
1
}
```

Thanks,
Noam

**Comment by sachsmc** [comment]
Hi again Noam,

You are totally right, in this case the ks do not factor. There is clearly some condition on the causal query that we are missing. I will continue thinking about it, but one possibly overly conservative condition is that the query must involve a counterfactual somewhere on the path from the left side to the outcome. Thus, I'm sorry to say that this method does not apply in your setting. 

I apologize, but thank you very much for raising the issue! I certainly did not anticipate this as a use-case of the package. 

Best wishes,
Mike

**Comment by nsfinkelstein** [comment]
Got it. Thank you for taking the time to discuss, and for your work on this!

Best,
Noam

**Comment by nsfinkelstein** [comment]
@sachsmc you may be interested to know that the same issue seems to arise for the graph below. For example, p{X(A=1)=1} and p{X(A=1)=0} return bounds that cannot be mutually sharp, in the same way that was described above.

This leads me to suspect that the following constraint

> the query must involve a counterfactual somewhere on the path from the left side to the outcome

is not strong enough. It seems to me that instead the query must involve only counterfactuals that consider intervention on (i) an element of the right side that blocks all paths between the left side and the outcome, or (ii) all elements of the left side that have open paths to the outcome.

![image](https://user-images.githubusercontent.com/2919482/82808429-fd770300-9e57-11ea-9afd-3e20e3159df0.png)

Thanks very much!

Best
Noam
